### PR TITLE
Shard `OpenedFileCache` to avoid lock contention

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Waiting on lock in `OpenedFileCache` could be noticeable sometimes. We sharded it into multiple sub-maps (each with its own lock) to avoid contention.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit 61a9d30d609df6ef4300b0d006d67b7f677b79cc with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51341/61a9d30d609df6ef4300b0d006d67b7f677b79cc/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/IO/OpenedFileCache.h [hunk 1]
```diff
@@ -26,50 +29,65 @@ namespace DB
...
+    static constexpr size_t buckets = 1024;
+    std::vector<OpenedFileMap> impls{buckets};
+
+public:
+    using OpenedFilePtr = OpenedFileMap::OpenedFilePtr;
+
+    OpenedFilePtr get(const std::string & path, int flags)
+    {
+        ProfileEventTimeIncrement<Microseconds> watch(ProfileEvents::OpenedFileCacheMicroseconds);
+        const auto bucket = sipHash64(path) % buckets;
```
Comment (alexey-milovidov) [inline src/IO/OpenedFileCache.h:89]: Ok, although even cityHash will suffice.

**Review by alexey-milovidov — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]

**Comment by filimonov** [comment]
One scenario when the issue can hurt you - is when you do FREEZE on s3 disk (that requires reading a lot of metadata files during hardlink creation). For example doing freeze on table with 7K parts can take like 40 minutes before the fix, most of the time waiting for those locks.

**Comment by alexey-milovidov** [comment]
FREEZE on S3 sounds like a bad idea. I hope you don't use "zero-copy replication", because it does not work with hard links at all.

**Comment by filimonov** [comment]
> I hope you don't use "zero-copy replication", because it does not work with hard links at all.

We don't. Actually it have some mechanics / cludges to mimic refcounts / hardlinks. But it's very fragile and unreliable. 

Awful thing here that FREEZE still creates frozen_metadata.txt and reads local metadata files no matter if you use 'zero-copy' or not. 

Here is the flamegraph of FREEZE command (based on trace_log) which was working for ~40 minutes.
flamegraph

P.S. @myrrc is working now on the alternative idea - you can take a look at the concept & work in progress here: https://github.com/ClickHouse/ClickHouse/pull/56269#issuecomment-1871009582 (still draft, and quite a lot of work to do)

**Pull request merged**